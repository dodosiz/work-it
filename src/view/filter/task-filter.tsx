import * as React from "react";
import { Button, Form, FormControl } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { removeFilter, setFilter } from "../../data/filter/filter";
import { usersSelector } from "../../data/users/users";

export function TaskFilter() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const [title, setTitle] = React.useState("");
	const [userId, setUserId] = React.useState<string | undefined>(undefined);
	const [dirty, setDirty] = React.useState(false);
	const handleAssigneeChange: React.ChangeEventHandler<HTMLSelectElement> = (
		event
	) => {
		setUserId(event.target.value);
		setDirty(true);
		if (!event.target.value && title === "") {
			dispatch(removeFilter());
		}
	};
	const handleTitleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setTitle(event.target.value);
		setDirty(true);
		if (event.target.value === "" && !userId) {
			dispatch(removeFilter());
		}
	};
	const handleFormSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		handleSubmit();
	};
	const handleSubmit = () => {
		dispatch(
			setFilter({
				taskTitle: title,
				userId,
			})
		);
		setDirty(false);
	};
	const isEmpty = !userId && title === "";
	return (
		<Form className="d-flex" onSubmit={handleFormSubmit}>
			<FormControl
				type="search"
				placeholder="Task title"
				className="me-2"
				aria-label="Search"
				value={title}
				data-testid="title-search"
				onChange={handleTitleChange}
			/>
			<Form.Select
				aria-label="Assignee"
				onChange={handleAssigneeChange}
				value={userId}
				className="me-2"
				data-testid="assignee-search"
			>
				<option></option>
				{users.map((user) => (
					<option key={user.id} value={user.id}>
						{user.firstName + " " + user.lastName}
					</option>
				))}
			</Form.Select>
			<Button
				onClick={handleSubmit}
				disabled={isEmpty || !dirty}
				variant="success"
			>
				Search
			</Button>
		</Form>
	);
}
