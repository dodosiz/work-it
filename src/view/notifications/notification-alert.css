.notification-alert {
	position: fixed;
	bottom: 1em;
	right: 1em;
	transition: all 250ms linear;
	width: 40%;
}
