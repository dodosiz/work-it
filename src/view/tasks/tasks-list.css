.task {
	box-shadow: rgba(0, 0, 0, 0.24) 0px 3px 8px;
	margin: 2em 0;
	opacity: 100%;
	transition: opacity 2s;
}

.pointer {
	cursor: pointer;
}

.done {
	color: green;
}

.undone {
	color: lightgreen;
}
