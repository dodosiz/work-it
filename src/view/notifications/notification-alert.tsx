import * as React from "react";
import { Alert, Fade } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	clearNotification,
	notificationSelector,
} from "../../data/notifications/notifications";
import "./notification-alert.css";

export function NotificationAlert() {
	const notificationTimeout = 2000;
	const dispatch = useDispatch();
	const notification = useSelector(notificationSelector);
	const show = !!notification;
	const fadeOut = () => {
		setTimeout(() => {
			dispatch(clearNotification());
		}, notificationTimeout);
	};
	const handleClose = () => {
		dispatch(clearNotification());
	};
	return (
		<Fade in={show} timeout={notificationTimeout} onEntered={() => fadeOut()}>
			<Alert
				className="notification-alert"
				dismissible
				show={show}
				variant="success"
				onClose={() => handleClose()}
			>
				{notification}
			</Alert>
		</Fade>
	);
}
