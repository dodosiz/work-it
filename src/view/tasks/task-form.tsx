import * as React from "react";
import { Button, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	clickedTaskSelector,
	closeTaskForm,
	createTask,
	taskFormModeSelector,
} from "../../data/tasks/tasks";
import { usersSelector } from "../../data/users/users";

export function TaskForm() {
	const dispatch = useDispatch();
	const users = useSelector(usersSelector);
	const clickedTask = useSelector(clickedTaskSelector);
	const readOnly = useSelector(taskFormModeSelector) === "readonly";
	const [title, setTitle] = React.useState(
		clickedTask ? clickedTask.title : ""
	);
	const [description, setDescription] = React.useState(
		clickedTask ? clickedTask?.description : ""
	);
	const [userId, setUserId] = React.useState<string | undefined>(
		clickedTask ? clickedTask.assignee?.id : undefined
	);
	const [dirty, setDirty] = React.useState(false);
	const handleTitleChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setTitle(event.target.value);
		setDirty(true);
	};
	const handleDescriptionChange: React.ChangeEventHandler<HTMLInputElement> = (
		event
	) => {
		setDescription(event.target.value);
		setDirty(true);
	};
	const handleAssigneeChange: React.ChangeEventHandler<HTMLSelectElement> = (
		event
	) => {
		setUserId(event.target.value);
		setDirty(true);
	};
	const handleSubmit: React.FormEventHandler<HTMLFormElement> = (event) => {
		event.preventDefault();
		if (!readOnly) {
			const assignee = users.find((user) => user.id === userId);
			dispatch(createTask({ title, description, assignee }));
			dispatch(
				addNotification({
					message: `Created task: ${title}`,
				})
			);
			dispatch(closeTaskForm());
		}
	};
	const handleCancel = () => {
		dispatch(closeTaskForm());
	};
	const submitDisabled = () => {
		return !isValid() || !dirty;
	};
	const isValid = () => {
		return title && userId;
	};
	return (
		<Form onSubmit={handleSubmit}>
			<Form.Group className="mb-3" controlId="title">
				<Form.Label>Title:</Form.Label>
				<Form.Control
					readOnly={readOnly}
					type="text"
					value={title}
					onChange={handleTitleChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="description">
				<Form.Label>Description:</Form.Label>
				<Form.Control
					type="textarea"
					readOnly={readOnly}
					value={description}
					onChange={handleDescriptionChange}
				/>
			</Form.Group>
			<Form.Group className="mb-3" controlId="assignee">
				<Form.Select
					aria-label="Assignee"
					onChange={handleAssigneeChange}
					disabled={readOnly}
					value={userId}
					data-testid="assignee"
				>
					<option></option>
					{users.map((user) => (
						<option key={user.id} value={user.id}>
							{user.firstName + " " + user.lastName}
						</option>
					))}
				</Form.Select>
			</Form.Group>
			{!readOnly && (
				<>
					<Button
						className="me-2"
						disabled={submitDisabled()}
						variant="primary"
						type="submit"
					>
						Save
					</Button>
					<Button variant="secondary" onClick={() => handleCancel()}>
						Cancel
					</Button>
				</>
			)}
		</Form>
	);
}
