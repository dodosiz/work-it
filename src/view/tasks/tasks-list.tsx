import * as React from "react";
import { Button, Card, Col, Row } from "react-bootstrap";
import {
	BsCheckCircle,
	BsCheckCircleFill,
	BsFillTrashFill,
	BsPencilFill,
} from "react-icons/bs";
import { useDispatch, useSelector } from "react-redux";
import { addNotification } from "../../data/notifications/notifications";
import {
	deleteTask,
	finishTask,
	openTaskForm,
	Task,
	tasksSelector,
} from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TasksList(props: { mode: "todo" | "done" }) {
	const tasks = useSelector(tasksSelector(props.mode));
	if (tasks.length === 0) {
		return (
			<p>
				{props.mode === "todo" ? "No tasks found." : "No finished tasks found."}
			</p>
		);
	}
	return (
		<>
			{tasks.sort(sortByDate).map((task) => (
				<TaskCard key={task.id} task={task} mode={props.mode} />
			))}
		</>
	);
}

function sortByDate(task1: Task, task2: Task) {
	if (task1.dateFinished && task2.dateFinished) {
		return task2.dateFinished.localeCompare(task1.dateFinished);
	} else {
		return 0;
	}
}

interface TaskCardProps {
	mode: "todo" | "done";
	task: Task;
}

function TaskCard(props: TaskCardProps) {
	const fadeOutTimeout = 1500; // a little faster then the CSS transition
	const [checked, setChecked] = React.useState(false);
	const dispatch = useDispatch();
	const handleCheck = () => {
		setChecked(true);
		setTimeout(() => {
			dispatch(finishTask({ taskId: props.task.id }));
		}, fadeOutTimeout);
	};
	const handleEdit = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "edit",
				clickedTask: props.task,
			})
		);
	};
	const handleDelete = () => {
		dispatch(deleteTask({ taskId: props.task.id }));
		dispatch(
			addNotification({
				message: `Deleted task "${props.task.title}".`,
			})
		);
	};
	const handleTitleClick = () => {
		dispatch(
			openTaskForm({
				taskFormMode: "readonly",
				clickedTask: props.task,
			})
		);
	};
	return (
		<Card className="task" style={checked ? { opacity: "0%" } : {}}>
			<Card.Body>
				<Row>
					<Col md={1}>
						<CheckBox
							taskId={props.task.id}
							checked={props.mode === "todo" ? checked : true}
							handleCheck={props.mode === "todo" ? handleCheck : undefined}
						/>
					</Col>
					<Col md={5}>
						<span
							className="pointer"
							data-testid={`title-${props.task.id}`}
							onClick={handleTitleClick}
						>
							{props.task.title}
						</span>
					</Col>
					<Col md={4}>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
					{props.mode === "done" && (
						<Col md={2}>{`Finished on: ${props.task.dateFinished}`}</Col>
					)}
					{props.mode === "todo" && (
						<Col md={2}>
							<Button
								data-testid={`edit-button-${props.task.id}`}
								variant="outline-primary"
								className="me-2"
								onClick={handleEdit}
							>
								<BsPencilFill />
							</Button>
							<Button
								data-testid={`delete-button-${props.task.id}`}
								variant="outline-danger"
								onClick={handleDelete}
							>
								<BsFillTrashFill />
							</Button>
						</Col>
					)}
				</Row>
			</Card.Body>
		</Card>
	);
}

interface CheckBoxProps {
	taskId: string;
	checked: boolean;
	handleCheck?: () => void;
}

function CheckBox(props: CheckBoxProps) {
	return (
		<div
			className={`${props.checked ? "done" : "undone"}`}
			data-testid={`check-${props.taskId}`}
		>
			{!props.checked && (
				<BsCheckCircle
					className="pointer"
					onClick={props.handleCheck}
					size={"1.5em"}
					data-testid={`check-box-${props.taskId}`}
				/>
			)}
			{props.checked && <BsCheckCircleFill size={"1.5em"} />}
		</div>
	);
}
