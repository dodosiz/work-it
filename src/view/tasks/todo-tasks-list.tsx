import * as React from "react";
import { Card, Col, Row } from "react-bootstrap";
import { Task } from "../../data/tasks/tasks";
import "./tasks-list.css";

export function TodoTasksList(props: { tasks: Task[] }) {
	if (props.tasks.length === 0) {
		return <p>No tasks to do. Start by creating a new task.</p>;
	}
	return (
		<>
			{props.tasks.map((task) => (
				<TodoTask key={task.id} task={task} />
			))}
		</>
	);
}

function TodoTask(props: { task: Task }) {
	return (
		<Card className="task">
			<Card.Body>
				<Row>
					<Col>{props.task.title}</Col>
					<Col>
						{props.task.assignee?.firstName +
							" " +
							props.task.assignee?.lastName}
					</Col>
				</Row>
			</Card.Body>
		</Card>
	);
}
