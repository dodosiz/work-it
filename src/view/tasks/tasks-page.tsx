import * as React from "react";
import { Container, Navbar, Nav, Button } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import {
	closeTaskForm,
	openTaskForm,
	taskFormModeSelector,
	taskFormOpenedSelector,
} from "../../data/tasks/tasks";
import { TaskFilter } from "../filter/task-filter";
import { ModalDialog } from "../modal/modal-dialog";
import { TaskForm } from "./task-form";
import { TasksList } from "./tasks-list";

export function TasksPage(props: { mode: "todo" | "done" }) {
	const dispatch = useDispatch();
	const taskFormOpened = useSelector(taskFormOpenedSelector);
	const taskFormMode = useSelector(taskFormModeSelector);
	const handleCloseTaskForm = () => {
		dispatch(closeTaskForm());
	};
	const handleCreateTaskClick = () => {
		dispatch(openTaskForm({ taskFormMode: "edit" }));
	};
	return (
		<>
			<ModalDialog
				handleClose={handleCloseTaskForm}
				show={taskFormOpened}
				title={taskFormMode === "readonly" ? "View task" : "Create new task"}
			>
				<TaskForm />
			</ModalDialog>
			<Navbar expand="lg" variant="light" bg="light">
				<Container>
					<Nav className="me-auto">
						{props.mode === "todo" && (
							<Button variant="success" onClick={handleCreateTaskClick}>
								Create Task
							</Button>
						)}
					</Nav>
					<TaskFilter />
				</Container>
			</Navbar>
			<Container>
				<TasksList mode={props.mode} />
			</Container>
		</>
	);
}
