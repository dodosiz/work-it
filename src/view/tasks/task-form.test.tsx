import * as React from "react";
import { configureStore } from "@reduxjs/toolkit";
import "uuid";
import { usersReducer } from "../../data/users/users";
import { notificationsReducer } from "../../data/notifications/notifications";
import { tasksReducer } from "../../data/tasks/tasks";
import { AppState } from "../../data/store";
import { Provider } from "react-redux";
import { TaskForm } from "./task-form";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { filterReducer } from "../../data/filter/filter";

jest.mock("uuid", () => ({
	v4: jest.fn().mockReturnValue("taskId"),
}));

const preloadedState: AppState = {
	notifications: { message: undefined },
	tasksState: { tasks: [], taskFormOpened: false, taskFormMode: "edit" },
	usersState: {
		users: [
			{
				id: "user1",
				firstName: "Alejandro",
				lastName: "Matthews",
				role: "Project manager",
			},
			{
				id: "user2",
				firstName: "Mai",
				lastName: "Yoder",
				role: "UI/UX designer",
			},
			{
				id: "user3",
				firstName: "Wade",
				lastName: "Delgado",
				role: "Marketing",
			},
		],
		userFormOpened: true,
		userIdToEdit: undefined,
	},
	filterState: {},
};

describe("Task form component:", () => {
	test("create a new task", () => {
		const store = configureStore({
			reducer: {
				usersState: usersReducer,
				tasksState: tasksReducer,
				notifications: notificationsReducer,
				filterState: filterReducer,
			},
			preloadedState,
		});
		render(
			<Provider store={store}>
				<TaskForm />
			</Provider>
		);
		userEvent.type(screen.getByLabelText("Title:"), "Create web app.");
		userEvent.type(
			screen.getByLabelText("Description:"),
			"Create a web app with React."
		);
		fireEvent.change(screen.getByTestId("assignee"), {
			target: { value: "user2" },
		});
		userEvent.click(screen.getByRole("button", { name: "Save" }));
		expect(store.getState().tasksState.tasks).toEqual([
			{
				id: "taskId",
				assignee: {
					firstName: "Mai",
					id: "user2",
					lastName: "Yoder",
					role: "UI/UX designer",
				},
				dateFinished: undefined,
				description: "Create a web app with React.",
				title: "Create web app.",
			},
		]);
		expect(store.getState().notifications.message).toBe(
			"Created task: Create web app."
		);
	});
	test("task opened in read only mode", () => {
		const store = configureStore({
			reducer: {
				usersState: usersReducer,
				tasksState: tasksReducer,
				notifications: notificationsReducer,
				filterState: filterReducer,
			},
			preloadedState: {
				...preloadedState,
				tasksState: {
					...preloadedState.tasksState,
					taskFormMode: "readonly",
					clickedTask: {
						id: "task1",
						assignee: {
							id: "user2",
							firstName: "John",
							lastName: "Doe",
							role: "Developer",
						},
						title: "Plan the requirements for the next release.",
						dateFinished: undefined,
						description: "",
					},
				},
			},
		});
		const editor = render(
			<Provider store={store}>
				<TaskForm />
			</Provider>
		);
		expect(editor.container).toMatchSnapshot();
	});
});
