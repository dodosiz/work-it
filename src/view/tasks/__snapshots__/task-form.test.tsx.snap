// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task form component: task opened in read only mode 1`] = `
<div>
  <form
    class=""
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="title"
      >
        Title:
      </label>
      <input
        class="form-control"
        id="title"
        readonly=""
        type="text"
        value="Plan the requirements for the next release."
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="description"
      >
        Description:
      </label>
      <input
        class="form-control"
        id="description"
        readonly=""
        type="textarea"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <select
        aria-label="Assignee"
        class="form-select"
        data-testid="assignee"
        disabled=""
        id="assignee"
      >
        <option />
        <option
          value="user1"
        >
          Alejandro Matthews
        </option>
        <option
          value="user2"
        >
          Mai Yoder
        </option>
        <option
          value="user3"
        >
          Wade Delgado
        </option>
      </select>
    </div>
  </form>
</div>
`;
